### DISCUSSION TITLE
Custom sort by distance from a lat/lng point

### DISCUSSION THREAD

**Discussion by will83** [directus/directus#2720] [Feature Requests]
I need to sort my results by distance based on a specified lat/lng couple provided in the query.
I have the sql statement to do that, it works well in PhpMyAdmin, but I can't find how to add a custom command to get this working.
I want to keep all the actual functionalities of the "items" endpoint, but add a custom "sort" query param :
`GET /:project/items/:collection?sort=distance&lat=37.421&lng=-122.084`

The SQL statement is : 
```
SELECT *, ( 6371 * acos( cos( radians($lat) ) * cos( radians( lat ) ) * cos( radians( lng ) - radians($lng) ) + sin( radians($lat) ) * sin(radians(lat)) ) ) AS distance 
FROM collection 
ORDER BY distance
```
where $lat and $lng are coming from the query and lat and lng are fields of my items in the specified :collection

Thank you for your help

**Comment by rijkvanzanten** [comment]
There's no way to add a custom parameter into an existing endpoint. You could create a custom endpoint, but like you mentioned, you'd lose the standard functionalities of the items endpoint.

It would be pretty cool to be able to sort on distances, but I'm not really sure how we would make that happen within the existing `sort` param 🤔 For example, `&sort=distance` will sort alphabetically on the value of the distance column, so that doesn't really fly.

I think sorting by distance is specialized enough to warrant a dedicated param, especially seeing the requirement for setting the "middle point" from which the distance is calculated

**Reply by will83** [reply]
What about `?distancefrom=lat,lng` param that override the sort param?
I really need that kind of sorting for my project. Is there a quick solution for my need instead of rewriting a custom endpoint?
If this is the only solution, is there an example somewhere i could rely on to make sql queries in my custom endpoint?

**Comment by reddo** [comment]
@rijkvanzanten pointed me to this discussion. I'm in a similar situation, I've posted a suggestion here: https://github.com/directus/api/issues/1916#issuecomment-635535567

I made a custom endpoint for this, that works. See here for the code: https://gist.github.com/reddo/0157a184d47e7f3f0caa7d253de432ff

Basically, what I'm doing is querying the DB to calculate the distances, then again with the itemsService to get all the items in the collection and merging the 2 arrays.

I would still prefer this to be a core feature.

I'm not at all familiar with ZenDB, sn't there a possibility to insert a query into the [findAll()](https://github.com/directus/api/blob/b07ecacf9f1ecc554c2cfa5a978cdd98540e6759/src/core/Directus/Services/ItemsService.php#L77https://github.com/directus/api/blob/b07ecacf9f1ecc554c2cfa5a978cdd98540e6759/src/core/Directus/Services/ItemsService.php#L77) function?

**Reply by rheymoss** [reply]
helpful, can you show me 'locations' structure which works with this endpoint?

**Reply by reddo** [reply]
I'm not sure, since I haven't worked with this for a while now, but I think it was the default location field from Directus with 

```
{
  lat: number;
  lng: number;
}
```
and apparently a `distance` query param (by looking at the code in the gist).

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
